scripts/RunNodes.hs

```diff
@@ -0,0 +1,128 @@
+   <> help "Full path to a chainweb-node binary")
+
+pNodes :: Parser Word8
+pNodes = option auto
+  (long "nodes" <> metavar "COUNT" <> value 10
+   <> help "Number of Nodes to run (default: 10)")
+
+pVersion :: Parser ChainwebVersion
+pVersion = option cver
+  (long "version" <> metavar "VERSION"
+   <> value (TestWithTime petersonChainGraph)
+   <> help "Chainweb Version to run the Nodes with (default: testWithTime)")
+  where
+    cver :: ReadM ChainwebVersion
+    cver = eitherReader $ \s ->
+        note "Illegal ChainwebVersion" . chainwebVersionFromText $ T.pack s
+
+pLogTarget :: Parser Log
+pLogTarget = option logPath
+  (long "log-target" <> metavar "local:PATH|es:ELASTICSEARCH_HOST:PORT"
+   <> help "Location to log to (default: STDOUT)")
+  where
+    logPath :: ReadM Log
+    logPath = eitherReader (A.parseOnly (esPath <|> plainPath) . T.pack)
+
+    esPath :: A.Parser Log
+    esPath = ES <$> host <*> port
+      where
+        host = A.string "es:" *> A.takeWhile1 (/= ':') <* A.char ':'
+        port = A.decimal
+
+    plainPath :: A.Parser Log
+    plainPath = Path <$> (A.string "local:" *> A.takeText)
+
+pLogLevel :: Parser T.Text
+pLogLevel = strOption
+  (long "log-level" <> metavar "info|warn|error|debug" <> value "info"
+   <> help "default: info")
+
+runNode :: Word8 -> Maybe T.Text -> Env -> IO ()
+runNode nid config (Env e ns v lt ll) = shelly $ run_ (fromText $ T.pack e) ops
+  where
+    logging :: Log -> [T.Text]
+    logging l@(ES _ _) =
+      [ sformat ("--telemetry-log-handle=" % stext) $ logText l
+      , sformat ("--log-handle=" % stext) $ logText l
+      ]
+    logging l@(Path _) =
+      [ sformat ("--telemetry-log-handle=" % stext % "/telemetry.node" % int % ".log") (logText l) nid
+      , sformat ("--log-handle=" % stext % "/node" % int % ".log") (logText l) nid
+      ]
+
+    logText :: Log -> T.Text
+    logText (ES h p) = sformat ("es:" % stext % ":" % int) h p
+    logText (Path p) = sformat ("file:" % stext) p
+
+    ops :: [T.Text]
+    ops = [ "--hostname=127.0.0.1"
+          , sformat ("--node-id=" % int) nid
+          , sformat ("--test-miners=" % int) ns
+          , sformat ("--chainweb-version=" % stext) $ plainVersionFromText v
+          , "--interface=127.0.0.1"
+          , sformat ("--log-level=" % stext) ll ]
+          <> maybe [] logging lt
+          <> maybe [] (\c -> [sformat ("--config-file=" % stext) c]) config
+          <> [ "+RTS", "-T" ]
+
+-- | Prep a local log dir for output.
+logDir :: Log -> IO ()
+logDir (ES _ _) = pure ()
+logDir (Path p) = shelly $ mkdir_p (fromText p)
+
+main :: IO ()
+main = do
+  env@(Env e ns _ lt _) <- execParser opts
+  print env
+  traverse_ logDir lt
+  whenM (doesFileExist e) $ do
+    -- Launch Bootstrap Node
+    boot <- async $ runNode 0 (Just "scripts/test-bootstrap-node.config") env
+    threadDelay 200000  -- 0.2 seconds
+    -- Launch Common Nodes
+    mapConcurrently_ (\n -> runNode n Nothing env) [1 .. ns - 1]
+    void $ wait boot
```

Comment:
If the bootstrap nodes didn't start or crashes, we wouldn't find out, because `mapConcurrently` blocks and we never get to this line. What about either running the bootstrap node along with the other nodes (using `mapConcurrently` or `runConcurrently`) or using something like `withAsync ... $ \boot -> link boot >> ...`?